File: examples/typescript/common/shutdown.ts
// Graceful SIGINT/SIGTERM handler for in-flight gRPC bidi streams.
//
// Lets the example clients tear down active streams cleanly on Ctrl-C
// instead of dumping a stack trace. Pulled into common/ so phone_call.ts
// + phone_call_burst.ts don't each re-implement the boilerplate.
//
// Usage:
//   const calls: Call[] = [...];
//   const shutdown = installSignalShutdown(calls);
//   const results = await Promise.allSettled(perCallPromises);
//   if (shutdown.seen) { ... } // signal-triggered, not natural completion
//
// Behaviour:
//   - First SIGINT/SIGTERM: prints a banner to stderr and calls .cancel()
//     on every still-open stream in the list.
//   - Subsequent signals: no-op (so a frantic Ctrl-C-Ctrl-C doesn't double-
//     cancel or re-print the banner).
//   - Captures the call list by reference at install time — calls added
//     later won't be cancelled. The example clients create all their
//     streams up front, so this is fine.

// We avoid importing the concrete grpc-js types here so the helper doesn't
// depend on @grpc/grpc-js. Any object with a `.cancel()` method works.
interface Cancellable {
  cancel(): void;
}

export interface Shutdown {
  seen: boolean;
}

export function installSignalShutdown(calls: Cancellable[]): Shutdown {
  const state: Shutdown = { seen: false };

  const request = (signame: string): void => {
    if (state.seen) return;
    state.seen = true;
    process.stderr.write(`\nReceived ${signame}, cancelling streams...\n`);
    for (const c of calls) {
      try { c.cancel(); } catch { /* best-effort */ }
    }
  };

  process.on("SIGINT", () => request("SIGINT"));
  process.on("SIGTERM", () => request("SIGTERM"));
  return state;
}

File: examples/python/common/shutdown.py
"""Graceful SIGINT/SIGTERM handler for asyncio task lists.

Lets the example clients tear down in-flight bidi streams cleanly on
Ctrl-C instead of unwinding through grpc.aio with a CancelledError
traceback. Pulled into common/ so phone_call.py + phone_call_burst.py
don't each re-implement the boilerplate.

Usage:
    tasks = [asyncio.create_task(...), ...]
    shutdown = install_signal_shutdown(tasks)
    results = await asyncio.gather(*tasks, return_exceptions=True)
    if shutdown.seen:
        ... # signal-triggered, not natural completion

Behaviour:
  - Handles SIGINT + SIGTERM identically.
  - First signal: prints "Received <signame>, cancelling streams..." to
    stderr and calls .cancel() on every still-running task.
  - Subsequent signals: no-op (so a frantic Ctrl-C-Ctrl-C doesn't double-
    cancel or re-print the banner).
  - Captures the task list by reference at install time — tasks added
    later won't be cancelled. The example clients create all their tasks
    up front, so this is fine.
"""
from __future__ import annotations

import asyncio
import signal
import sys
from collections.abc import Iterable
from dataclasses import dataclass


@dataclass
class Shutdown:
    """Mutable flag the signal handler flips. Poll `.seen` after gather()."""
    seen: bool = False


def install_signal_shutdown(tasks: Iterable[asyncio.Task]) -> Shutdown:
    """Install SIGINT + SIGTERM handlers that cancel every task in `tasks`.

    Returns a `Shutdown` whose `.seen` flag flips the first time a signal
    fires. Callers use it to distinguish 'cancelled by user' from
    'completed normally' when shaping the final summary.
    """
    state = Shutdown()
    task_list = list(tasks)

    def _request_shutdown(signame: str) -> None:
        if state.seen:
            return
        state.seen = True
        print(f"\nReceived {signame}, cancelling streams...", file=sys.stderr, flush=True)
        for t in task_list:
            if not t.done():
                t.cancel()

    loop = asyncio.get_running_loop()
    for sig, name in ((signal.SIGINT, "SIGINT"), (signal.SIGTERM, "SIGTERM")):
        loop.add_signal_handler(sig, _request_shutdown, name)
    return state
